Add interface guidelines for underline nav (#458)
primer/design: content/components/underline-nav.mdx

---
title: Underline nav
description: The UnderlineNav is used to display primary navigation.
componentId: underline_nav
reactId: underline_nav
figma: https://www.figma.com/file/GCvY3Qv8czRgZgvl1dG6lp/Primer-Web?node-id=18843-67449&t=XqqSolC6AzRO9nae-11
rails: https://primer.style/view-components/components/alpha/underlinenav/
---

# Underline nav

> The underline nav is used to provide navigation links to related views for the user's current context. It is typically used as primary navigation at the top of the page or as secondary navigation when nested in views with a [nav list](/components/nav-list).

Reviewer: Non-blocking nitpick: "at the top of the page" could be misleading. In dotcom, the top of the page has the global GitHub header bar and navigation.

## Anatomy

![Anatomy of the UnderlineNav component](https://user-images.githubusercontent.com/175638/186432408-c2479dd2-31fe-4261-9f9a-e67d49b28e72.png)

1. **Item**: A navigation link.
2. **Current item**: The current selection item is highlighted with an underline.
3. **Label**: Describes the navigation item.
4. **Leading icon (optional)**: An icon that identifies the navigation item.
5. **Counter (optional)**: Displays a count next to the label.

> 6. **Overflow menu**: A disclosure menu that displays additional items when there are too many to fit in the available space.

Reviewer: I don't think these need to be numbered. I'd say either convert it to an unordered list, or make each line it's own `<p>` by putting a return between each line.

## Content

### Item

The underline nav is composed of a list of items. Each item is a link that can be selected to navigate to a different page in the current context.

#### Label text

> The label indicates the destination page of the link. Keep labels clear, short, and concise.

Reviewer: "destination page" makes it sound like the user will leave their current context when clicking the link.

Maybe something like this?

```suggestion
The title of the view associated with that tab. Keep labels clear, short, and concise.
```

#### Leading icons

Use an Octicon to improve scannability of common items, but only when relevant. When adding leading icons, all items in the navigation should have one.

Make sure that icons used are consistent across views, and that they are not repeated for different navigations across the product.

#### Counters

You may include a counter to indicate the number of related resources contained in the desitination page, such as the number of open issues.

When loading multiple counters asynchronously, wait for all the data to be ready before displaying the counters, so you can avoid multiple layout shifts. Use the counter loading state while waiting for the data.

## Layout

### Placement

The underline nav has two possible main placements depending on its intended usage:

- As primary navigation: place it at the top of the page. Ensure any secondary and tertiary navigation is nested under the underline nav.

> - As secondary or tertiary navigation: place it above the main content of the.

Reviewer: I think you forgot to finish this sentence 😅

In either scenarios, avoid stacking multiple uderline navs to display multiple levels of navigation.

### Alignment

> Align the the underline nav to the left and, ehen possible, make the component span the available width.

Reviewer: ```suggestion
Align the the underline nav to the left and, when possible, make the component span the available width.
```

### Responsive design

> Underline nav handles overflowing items automatically by collapsing them into a disclosure menu. When some of the items don't fit the available space, the component will first hide leading icons to economize space and then move items to the overflow menu one by one.

Reviewer: It might be nice to add a visual example above this paragraph.

Author: Yes! I've added it.

> ## Accessibility

Reviewer: I'm kind of surprised we don't treat this component semantically like tabs. I assume it's because of the overflow behavior?

Reviewer: I talked to a11y about this, and I understand why we don't treat this semantically like tabs.

Set `aria-current` to `"page"` to indicate that the item represents the current page. Set `aria-current` to `"location"` to indicate that the item represents the current location on a page.

### Keyboard navigation

Items can be activated using a cursor or keyboard

| Key(s)                                                                                                                                                    | Description                   |
| --------------------------------------------------------------------------------------------------------------------------------------------------------- | ----------------------------- |
| <Box as="kbd" backgroundColor="canvas.subtle" borderRadius={1} borderColor="border.default" borderWidth={1} borderStyle="solid" py={1} px={2}>Enter</Box> | Activates the focused item.   |
| <Box as="kbd" backgroundColor="canvas.subtle" borderRadius={1} borderColor="border.default" borderWidth={1} borderStyle="solid" py={1} px={2}>Tab</Box>   | Moves focus to the next item. |

## Related links

> - [TabNav](/components/tab-nav)

Reviewer: ```suggestion
- [Tab nav](/components/tab-nav)
```

> - [Nav list](/components/nav-list)

Reviewer: Non-blocking nitpick: maybe we could alphabetize these?

---
title: Underline nav
description: The UnderlineNav is used to display navigation.
componentId: underline_nav
reactId: underline_nav

> figma: https://www.figma.com/file/GCvY3Qv8czRgZgvl1dG6lp/Primer-Web?node-id=18843-67449&t=XqqSolC6AzRO9nae-11

Reviewer 2: we could add a `figmaId:  underline_nav` instead of the figma URL, so merging this with the figma automation docs will avoid any possible conflict

export default ComponentLayout

## Usage

The underline nav is used to provide navigation links to related views for the user's current context. It is typically used as primary navigation at the top of the page or as secondary or tertiary navigation when nested in views with a [nav list](/components/nav-list).

## Anatomy

![Anatomy of the UnderlineNav component](https://user-images.githubusercontent.com/175638/186432408-c2479dd2-31fe-4261-9f9a-e67d49b28e72.png)

- **Item**: A navigation link.
- **Current item**: The item currently selected is highlighted with an underline.
- **Label**: Describes the navigation item.
- **Leading icon (optional)**: An icon that identifies the navigation item.
- **Counter (optional)**: Displays a count next to the label.
- **Overflow menu**: A disclosure menu that displays additional items when there are too many to fit in the available space.

## Options

### Item

The underline nav is composed of a list of items. Each item is a link that can be selected to navigate to a different page in the current context.

#### Label text

The title of the view associated with that tab. Keep labels clear, short, and concise.

#### Leading icons

> Use an Octicon to improve scannability of common items, but only when relevant. When adding leading icons, all items in the navigation should have one.

Reviewer 2: Maybe adding a link to Octicons? 

```suggestion
Use an [Octicon](/foundations/icons) to improve the scannability of common items, but only when relevant. When adding leading icons, all items in the navigation should have one.
```
